language/move-core/types/src/metadata.rs

```diff
@@ -0,0 +1,11 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+/// Representation of metadata,
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Metadata {
+    /// The key identifying the type of metadata.
+    pub key: Vec<u8>,
+    /// The value of the metadata.
+    pub value: Vec<u8>,
+}
```

Comment:
I'm wonder if we should make the in-memory Metadata representation more typed. For example,
```
pub enum Metadata {
    CompilerVersion(String),
}
```
or maybe
```
pub struct Metadata {
    compiler_version: Option<String>,
    ...
}
// now CompiledModule and CompiledScript would just own a Metadata instead of Vec<Metadata>
```
Brining this up since a lot other stuff in the file format's in-memory representation are typed, so it might make sense to stick to that style. Could I ask if you would want to allow users to inject custom metadata one day?

Reply:
I think having the key/value metadata representation is better, as it allows to build a key/value map from this representation which can be quickly accessed if only part of the metadata is needed. Also some of the metadata in this map may not be even known to a particular configuration of the VM, so it doesn't have the related types in the binary.